File: packages/swift-sdk/SwiftExampleApp/SwiftExampleApp/Services/AddressFundFromAssetLockCoordinator.swift
import Combine
import Foundation
import SwiftDashSDK

/// Singleton hub for in-flight platform-address funding attempts,
/// hosted on `PlatformWalletManager` so funds survive view
/// dismissal and network-toggle pressure.
///
/// Mirrors [`RegistrationCoordinator`] for the
/// `AddressFundingFromAssetLockTransition` flow. Keyed by
/// `(walletId, platformAccountIndex, recipientHash)` — that's the
/// natural unit of work since a wallet can fund many addresses
/// concurrently and "next unused" address allocation happens
/// Rust-side per call. The single-flight invariant prevents a user
/// from double-tapping the same address-funding submission during
/// the asset-lock broadcast window.
@MainActor
final class AddressFundFromAssetLockCoordinator: ObservableObject {
    /// Composite key — needs `Hashable` so the map can index by it.
    /// `walletId` is 32 raw bytes; `recipientHash` is 20 raw bytes;
    /// `platformAccountIndex` is the DIP-17 account that owns the
    /// recipient address.
    struct SlotKey: Hashable {
        let walletId: Data
        let platformAccountIndex: UInt32
        let recipientHash: Data
    }

    /// Active controllers keyed by slot. Stored as `@Published` so
    /// the "Pending Platform Top Ups" row on the Wallet Detail
    /// screen can observe map mutations via `objectWillChange`.
    @Published private(set) var controllers: [SlotKey: AddressFundFromAssetLockController] = [:]

    /// Per-slot subscriptions that forward each controller's
    /// `objectWillChange` (fired on a `phase` transition) to this
    /// coordinator's own `objectWillChange`.
    ///
    /// The `@Published controllers` map only re-publishes on
    /// insert/remove, NOT when a controller already in the map flips
    /// phase (`.inFlight → .failed`/`.completed`, or a `.failed`-slot
    /// retry that reuses the controller). A view that observes only the
    /// coordinator — e.g. the wallet-detail "Pending Top Ups" list, whose
    /// Resume gate reads controller `phase` — would otherwise render from
    /// stale phase state across those transitions. Keyed by slot so the
    /// subscription is torn down with the controller it tracks.
    private var phaseObservers: [SlotKey: AnyCancellable] = [:]

    /// True when at least one slot is currently in flight (phase
    /// `.inFlight`). Used by the network toggle's `.disabled(_:)`
    /// modifier — switching testnet ↔ mainnet mid-flight tears down
    /// the FFI manager and would abort the in-flight call.
    var hasInFlightFundings: Bool {
        controllers.contains { _, controller in
            if case .inFlight = controller.phase { return true }
            return false
        }
    }

    /// Look up the controller for a slot if one exists. Returns
    /// `nil` when there's no active funding for the slot — callers
    /// use that to decide whether to spawn a new controller or
    /// reuse the existing one.
    func controller(
        walletId: Data,
        platformAccountIndex: UInt32,
        recipientHash: Data
    ) -> AddressFundFromAssetLockController? {
        controllers[
            SlotKey(
                walletId: walletId,
                platformAccountIndex: platformAccountIndex,
                recipientHash: recipientHash
            )
        ]
    }

    /// Snapshot of every active controller, sorted by recency of
    /// last submit (most recent first). Used by the "Pending
    /// Platform Funding" row so dismissed-but-still-running flows
    /// remain reachable.
    func activeControllers() -> [AddressFundFromAssetLockController] {
        controllers.values.sorted { lhs, rhs in
            (lhs.lastSubmittedAt ?? .distantPast) > (rhs.lastSubmittedAt ?? .distantPast)
        }
    }

    /// Start a funding for the slot, or reuse an existing
    /// controller if one is already in flight for it. Returns the
    /// controller for `FundFromAssetLockPlatformAddressView` to bind a
    /// `AddressFundFromAssetLockProgressView` against.
    ///
    /// Single-flighting is enforced here at the coordinator level
    /// because the controller's `submit()` only guards within its
    /// own phase machine — without a phase check before fresh-slot
    /// creation, a second tap during the FFI window would race two
    /// FFI calls for the same asset lock.
    func startFunding(
        walletId: Data,
        platformAccountIndex: UInt32,
        recipientHash: Data,
        recipientType: UInt8,
        body: @escaping () async throws -> UInt64
    ) -> AddressFundFromAssetLockController {
        let key = SlotKey(
            walletId: walletId,
            platformAccountIndex: platformAccountIndex,
            recipientHash: recipientHash
        )
        if let existing = controllers[key] {
            switch existing.phase {
            case .inFlight, .completed:
                // Active or just-completed — don't re-enter. Returning
                // the existing controller lets the caller bind to its
                // progress / terminal state without disrupting it.
                return existing
            case .idle, .failed:
                // Legitimate restart paths.
                existing.submit(body: body)
                // No retention sweep here — the slot is sticky on
                // .failed (we want the user to see + dismiss the
                // error) and a duplicate sweep on retry would just
                // spawn a second 30s poll Task against the same
                // controller. Sweep was already scheduled when the
                // controller was first created.
                return existing
            }
        }
        let controller = AddressFundFromAssetLockController(
            walletId: walletId,
            platformAccountIndex: platformAccountIndex,
            recipientHash: recipientHash,
            recipientType: recipientType
        )
        controllers[key] = controller
        // Forward this controller's phase transitions to observers of the
        // coordinator (see `phaseObservers`). Registered before `submit()`
        // so the initial `.idle → .inFlight` flip re-publishes too.
        phaseObservers[key] = controller.objectWillChange.sink { [weak self] _ in
            self?.objectWillChange.send()
        }
        controller.submit(body: body)
        scheduleRetentionSweep(key: key, controller: controller)
        return controller
    }

    /// Manually drop a controller from the map. Used by the UI's
    /// "Dismiss" action on a `.failed` row (failures stay
    /// indefinitely until acknowledged so the user can read the
    /// error).
    func dismiss(
        walletId: Data,
        platformAccountIndex: UInt32,
        recipientHash: Data
    ) {
        let key = SlotKey(
            walletId: walletId,
            platformAccountIndex: platformAccountIndex,
            recipientHash: recipientHash
        )
        controllers.removeValue(forKey: key)
        phaseObservers.removeValue(forKey: key)
    }

    // MARK: - Retention sweep

    /// Auto-purge `.completed` controllers ~30s after the success
    /// transition so the wallet's Pending list doesn't accumulate
    /// stale rows. `.failed` controllers stay indefinitely until
    /// the user dismisses them. Same shape as
    /// `RegistrationCoordinator`.
    private func scheduleRetentionSweep(
        key: SlotKey,
        controller: AddressFundFromAssetLockController
    ) {
        Task { [weak self, weak controller] in
            guard let controller = controller else { return }
            var completedAt: Date?
            while !Task.isCancelled {
                let phase = await MainActor.run { controller.phase }
                switch phase {
                case .completed:
                    if completedAt == nil {
                        completedAt = Date()
                    } else if let at = completedAt,
                              Date().timeIntervalSince(at) >= 30 {
                        await MainActor.run {
                            _ = self?.controllers.removeValue(forKey: key)
                            self?.phaseObservers.removeValue(forKey: key)
                        }
                        return
                    }
                case .failed:
                    return
                default:
                    completedAt = nil
                }
                try? await Task.sleep(nanoseconds: 1_000_000_000)
            }
        }
    }
}

File: packages/kotlin-sdk/KotlinExampleApp/app/src/main/java/org/dashfoundation/example/services/assetlock/AddressFundFromAssetLockCoordinator.kt
package org.dashfoundation.example.services.assetlock

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * Hub for in-flight "fund a Platform address from an asset lock" attempts
 * — 1:1 port of `AddressFundFromAssetLockCoordinator.swift`. Held on the
 * app container (the Kotlin analog of hosting it on `PlatformWalletManager`)
 * so fundings survive screen dismissal and network-toggle pressure.
 *
 * Keyed by `(walletId, platformAccountIndex, recipientHash)` — a wallet can
 * fund many addresses concurrently on the same account, so the slot key is
 * composite. [startFunding] single-flights on the slot key, and
 * [hasInFlightFundings] gates the Settings network picker alongside the
 * registration coordinator.
 *
 * Mirrors [org.dashfoundation.example.services.RegistrationCoordinator]:
 * state lives behind [controllers] (a [StateFlow] map) so Compose observers
 * pick up mutations; the retention sweep runs on [scope] so `.Completed`
 * rows auto-purge ~[retentionMillis] after success. [retentionMillis] /
 * [pollMillis] / [now] are injectable so tests drive the sweep on virtual
 * time.
 */
class AddressFundFromAssetLockCoordinator(
    private val scope: CoroutineScope,
    private val retentionMillis: Long = 30_000L,
    private val pollMillis: Long = 1_000L,
    private val now: () -> Long = System::currentTimeMillis,
) {
    /** Composite slot key. Binary ids compared as hex. ← Swift `SlotKey`. */
    data class SlotKey(
        val walletIdHex: String,
        val platformAccountIndex: Int,
        val recipientHashHex: String,
    )

    private val _controllers =
        MutableStateFlow<Map<SlotKey, AddressFundFromAssetLockController>>(emptyMap())

    /** Active controllers keyed by slot, observable by the pending-fundings list. */
    val controllers: StateFlow<Map<SlotKey, AddressFundFromAssetLockController>> =
        _controllers.asStateFlow()

    /**
     * True when at least one slot is [AddressFundFromAssetLockController.Phase.InFlight]
     * — used by the network picker's enabled gate. ← Swift `hasInFlightFundings`.
     */
    val hasInFlightFundings: Boolean
        get() = _controllers.value.values.any { it.phase.value.isActive }

    /** The controller for a slot, or null if none is active. ← Swift `controller(...)`. */
    fun controller(
        walletId: ByteArray,
        platformAccountIndex: Int,
        recipientHash: ByteArray,
    ): AddressFundFromAssetLockController? =
        _controllers.value[key(walletId, platformAccountIndex, recipientHash)]

    /**
     * Every active controller, most-recently-submitted first — the order
     * the pending-fundings list renders. ← Swift `activeControllers()`.
     */
    fun activeControllers(): List<AddressFundFromAssetLockController> =
        _controllers.value.values.sortedByDescending { it.lastSubmittedAt ?: Long.MIN_VALUE }

    /**
     * Start a funding for the slot, or reuse the existing controller if one
     * is already in flight. Single-flighting is enforced HERE: a second tap
     * during the FFI window returns the existing controller and does NOT
     * re-fire the body. [Phase.Idle] / [Phase.Failed] are legitimate
     * restarts. ← Swift `startFunding`.
     */
    fun startFunding(
        walletId: ByteArray,
        platformAccountIndex: Int,
        recipientHash: ByteArray,
        recipientType: Int,
        body: suspend () -> Long,
    ): AddressFundFromAssetLockController {
        val key = key(walletId, platformAccountIndex, recipientHash)
        val existing = _controllers.value[key]
        if (existing != null) {
            when (existing.phase.value) {
                is AddressFundFromAssetLockController.Phase.InFlight,
                is AddressFundFromAssetLockController.Phase.Completed,
                -> return existing // active / just-completed — don't re-enter
                is AddressFundFromAssetLockController.Phase.Idle,
                is AddressFundFromAssetLockController.Phase.Failed,
                -> {
                    existing.submit(body)
                    scheduleRetentionSweep(key, existing)
                    return existing
                }
            }
        }
        val controller = AddressFundFromAssetLockController(
            walletId,
            platformAccountIndex,
            recipientHash,
            recipientType,
            scope,
            now,
        )
        _controllers.update { it + (key to controller) }
        controller.submit(body)
        scheduleRetentionSweep(key, controller)
        return controller
    }

    /**
     * Manually drop a controller from the map — the UI's "Dismiss" action
     * on a [Phase.Failed] row. ← Swift `dismiss`.
     */
    fun dismiss(walletId: ByteArray, platformAccountIndex: Int, recipientHash: ByteArray) {
        _controllers.update { it - key(walletId, platformAccountIndex, recipientHash) }
    }

    /**
     * Auto-purge a [Phase.Completed] controller ~[retentionMillis] after
     * the success transition so the pending list doesn't accumulate stale
     * rows. [Phase.Failed] stays indefinitely until the user dismisses it.
     * Polls every [pollMillis]. ← Swift `scheduleRetentionSweep`.
     */
    private fun scheduleRetentionSweep(
        key: SlotKey,
        controller: AddressFundFromAssetLockController,
    ) {
        scope.launch {
            var completedAt: Long? = null
            while (true) {
                when (controller.phase.value) {
                    is AddressFundFromAssetLockController.Phase.Completed -> {
                        val nowMs = now()
                        if (completedAt == null) {
                            completedAt = nowMs
                        } else if (nowMs - completedAt >= retentionMillis) {
                            _controllers.update { it - key }
                            return@launch
                        }
                    }
                    is AddressFundFromAssetLockController.Phase.Failed -> return@launch
                    else -> completedAt = null
                }
                delay(pollMillis)
            }
        }
    }

    private fun key(
        walletId: ByteArray,
        platformAccountIndex: Int,
        recipientHash: ByteArray,
    ): SlotKey = SlotKey(walletId.toHex(), platformAccountIndex, recipientHash.toHex())

    private fun ByteArray.toHex(): String = joinToString("") { "%02x".format(it) }
}
